File: backend/internal/modules/bjj/proficiency.go
package bjj

import (
	"context"
	"time"
)

// Proficiency is one technique's accumulated evidence.
//
// NOT A SCORE, and deliberately not. `docs/decisions/bjj-tracking-design.md`
// rules out asking anyone to rate their triangle 1–5 — people are bad at it,
// it goes stale, and it produces a number with no provenance. What this
// returns is the facts: drilled it twelve times, went for it three times in
// rolling, landed it twice. A reader can form a judgement from that and can
// always see what the judgement rests on.
type Proficiency struct {
	TechniqueID string `json:"technique_id"`
	// Name/Position/Category come from the shared library. They are here so a
	// client can render the list without a second fetch and a join of its own.
	Name     string `json:"name"`
	Position string `json:"position"`
	Category string `json:"category"`

	// The funnel. Drilled is practice; Attempted is went-for-it-and-missed;
	// Scored is landed it live. ATTEMPTED AND SCORED ARE DISJOINT — see the
	// migration — so Attempted+Scored is how often it was tried, and the
	// drilled→tried drop-off is the number the whole design is built around.
	Drilled   int `json:"drilled"`
	Attempted int `json:"attempted"`
	Scored    int `json:"scored"`
	// Conceded is this technique done TO the athlete. Nothing writes it today
	// (no client can author a technique-tagged conceded row) but the API
	// accepts one, so it is reported rather than silently dropped.
	Conceded int `json:"conceded"`
	// Defended is them going for it and the athlete stopping them — the
	// mirror of Attempted, and the half a roadmap's defensive criterion is
	// counted from. Reported alongside Conceded so the pair reads as one
	// exchange seen from both ends rather than as two unrelated tallies.
	Defended int `json:"defended"`

	// Sessions is how many separate sessions contributed, which is the
	// honesty check on every number above: twelve reps in one class is not
	// the same evidence as twelve reps across six weeks.
	Sessions int       `json:"sessions"`
	LastSeen time.Time `json:"last_seen"`
}

// Tried is how often the technique was attempted live at all.
func (p Proficiency) Tried() int { return p.Attempted + p.Scored }

// ProficiencySummary is the top-level funnel — the three numbers that are
// actually actionable before you look at any individual technique.
type ProficiencySummary struct {
	// Techniques is how many distinct techniques have any evidence at all.
	Techniques int `json:"techniques"`
	// Drilled is how many were ever drilled; TriedLive how many of those were
	// ever taken into a live round; Landed how many ever worked.
	//
	// These are COUNTS OF TECHNIQUES, not counts of reps. "You have drilled 34
	// techniques and taken 6 of them into a roll" is a finding. "You have done
	// 210 reps" is a statistic.
	Drilled   int `json:"drilled"`
	TriedLive int `json:"tried_live"`
	Landed    int `json:"landed"`
}

// SummariseProficiency folds the rows into the headline funnel.
//
// A pure function over the list rather than a second aggregate query: it
// cannot disagree with the rows the client is shown, which a separate
// COUNT(*) with its own WHERE clause eventually would. It is also the half
// worth testing, and this way it is testable without a database.
func SummariseProficiency(rows []Proficiency) ProficiencySummary {
	var s ProficiencySummary
	for _, p := range rows {
		s.Techniques++
		if p.Drilled > 0 {
			s.Drilled++
		}
		if p.Tried() > 0 {
			s.TriedLive++
		}
		if p.Scored > 0 {
			s.Landed++
		}
	}
	return s
}

// maxProficiencyRows bounds the list.
//
// Every list endpoint has a ceiling — `apihttp.ConditionalGet` buffers the
// whole response body to hash it, so peak memory per in-flight request is
// bounded by the largest response the API can produce, and that is only true
// if nothing is unbounded.
//
// It cannot bind today: the GROUP BY is on technique_id, which has an FK to
// `techniques`, so the row count is capped by the library — 466 entries. A
// client CANNOT reach this by inventing ids, as an earlier version of this
// comment claimed; the FK rejects them as invalid input.
//
// The only way it ever binds is the LIBRARY growing past 500, at which point
// the funnel starts truncating silently — no pagination, no error, and the
// summary folds from the truncated rows so it under-reports in step. That is
// pinned by a test asserting the catalog stays under this number, which is the
// version of this guard that can actually fail.
const maxProficiencyRows = 500

// ProficiencyRepository is the read side of the technique funnel.
type ProficiencyRepository interface {
	// ListProficiency returns one row per technique the athlete has evidence
	// for, most-evidence first.
	ListProficiency(ctx context.Context, userID string) ([]Proficiency, error)
}

File: apps/mobile/lib/proficiency.ts
import { apiRequest } from './apiRequest';
import type { TokenGetter } from './useAuthToken';

/**
 * The technique funnel, aggregated across every session — client half of
 * `GET /v1/bjj/proficiency`. (The path below omits the prefix: `apiRequest`
 * already carries `/v1`, and repeating it here would request `/v1/v1/...`.)
 *
 * **`attempted` and `scored` are DISJOINT**, which the backend's own comment
 * insists on and which any reader here has to carry: `attempted` is "went for
 * it and it did not land", not "total tries". So `attempted + scored` is how
 * often it was tried, and `scored / (attempted + scored)` is the hit rate.
 * Reading `attempted` as total tries is the natural mistake and gives
 * different numbers — see `lib/suggestion.ts`, where getting it wrong would
 * tell an athlete to try something they are already landing.
 *
 * `apps/web` has its own copy of this type. Not shared, because the two apps
 * cannot import from each other and there is no `packages/` yet — the same gap
 * the adherence rule ran into. Worth a shared package the third time this
 * happens.
 */
export type Proficiency = {
  technique_id: string;
  name: string;
  position: string;
  category: string;
  drilled: number;
  attempted: number;
  scored: number;
  conceded: number;
  /**
   * Them going for it and the athlete stopping them — the mirror of
   * `attempted`, and the defensive half of the 2x2 the tag vocabulary forms.
   * Added for N84: `bucketOf` below reads it, and its absence is exactly what
   * made `apps/web`'s own proficiency page briefly mis-bucket a pure-defence
   * technique as "used on you" — see that page's own note. Required by the
   * contract, matching `apps/web/src/lib/api.ts`'s `BjjProficiency`.
   */
  defended: number;
  /** How many separate sessions contributed — the honesty check on the rest. */
  sessions: number;
  last_seen: string;
};

/** Counts of TECHNIQUES, not of reps — matches `apps/web`'s `BjjProficiencySummary`. */
export type ProficiencySummary = {
  techniques: number;
  drilled: number;
  tried_live: number;
  landed: number;
};

/**
 * Read the funnel, with the headline it folds to.
 *
 * Network-only, deliberately — see {@link fetchProficiency}'s note, which
 * applies here unchanged. This is the full response; `fetchProficiency` below
 * is a thin projection of it kept for its existing callers (the Today card's
 * suggestion logic and the reflection wizard), which want the list and have
 * never wanted the summary. Both go through ONE request rather than two
 * disagreeing reads of the same endpoint.
 */
export function fetchProficiencyFull(
  getToken: TokenGetter,
  signal?: AbortSignal,
): Promise<{ techniques: Proficiency[]; summary: ProficiencySummary | null }> {
  // `?? []` / `?? null` at the parse boundary, for the reason `bjjFocus.ts`
  // documents: a drifted or older server omitting a field must not hand
  // `undefined` to a consumer inside a `useMemo`, which takes the render down
  // rather than degrading gracefully.
  return apiRequest<{ techniques?: Proficiency[]; summary?: ProficiencySummary }>(
    getToken,
    '/bjj/proficiency',
    { signal },
  ).then((r) => ({ techniques: r.techniques ?? [], summary: r.summary ?? null }));
}

/**
 * Read the funnel.
 *
 * Network-only, deliberately. This is an aggregate over every session the
 * athlete has ever logged, including ones synced from another device, so a
 * local answer would be a different and quietly smaller number. The caller
 * treats a failure as "no suggestion", never as an error worth a banner:
 * a suggestion is an offer, and an offer that cannot be made is not a fault
 * the athlete needs telling about.
 */
export function fetchProficiency(
  getToken: TokenGetter,
  signal?: AbortSignal,
): Promise<Proficiency[]> {
  return fetchProficiencyFull(getToken, signal).then((r) => r.techniques);
}

/** Below this, a hit rate is noise rather than a measurement. Matches web's page. */
export const MIN_TRIES_FOR_RATE = 5;

export type Bucket = 'all' | 'untried' | 'working' | 'stalled' | 'against';

/**
 * Every row lands in exactly one bucket, and the chip counts sum to
 * "Everything" — a straight port of `apps/web/src/app/dashboard/proficiency/page.tsx`'s
 * `bucketOf`, kept byte-for-byte in its reasoning because the two screens
 * describe the same funnel and must not disagree about which bucket a
 * technique falls in.
 *
 * `against` is reachable and returns rows the API's own filter allows through
 * (`technique_id IS NOT NULL`, nothing more) — a technique whose only evidence
 * is a `conceded` tag comes back with zeroes across drilled/attempted/scored
 * and `defended`. No shipped client authors one today; it still has to be
 * handled rather than assumed away, or the chip counts silently stop summing
 * to "Everything".
 */
export function bucketOf(p: Proficiency): Exclude<Bucket, 'all'> {
  const tried = p.attempted + p.scored;
  if (tried > 0) return p.scored > 0 ? 'working' : 'stalled';
  if (p.drilled > 0) return 'untried';
  // Defending it IS live evidence, and of the athlete succeeding — see web's
  // note on why this is checked before falling through to `against`.
  if (p.defended > 0) return 'working';
  return 'against';
}
